moveit_ros/moveit_servo/include/moveit_servo/servo_parameters.cpp

```diff
@@ -0,0 +1,278 @@
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author    : Adam Pettinger
+   Desc      : Declares, loads, and checks ServoParameters for Servo
+   Title     : servo_parameters.cpp
+   Project   : moveit_servo
+   Created   : 07/02/2020
+*/
+
+
+#include <rclcpp/rclcpp.hpp>
+
+#include <type_traits>
+
+#include <moveit_servo/servo_parameters.h>
+
+namespace moveit_servo
+{
+template <typename T>
+bool declareAndGetParam(const std::string& param_name, T& output_value, rclcpp::Node::SharedPtr& node,
+                        const rclcpp::Logger& logger)
+{
+  bool got_param = false;
+  try
+  {
+    node->declare_parameter<T>(param_name, T{});
+    got_param = node->get_parameter(param_name, output_value);
+  }
+  catch(const rclcpp::exceptions::InvalidParameterTypeException& e)
+  {
+    // Catch a <double> parameter written in the yaml as "1" being considered an <int>
+    if(std::is_same<T, double>::value)
+    {
+      int value;
+      node->undeclare_parameter(param_name);
+      node->declare_parameter<int>(param_name, 0);
+      got_param = node->get_parameter(param_name, value);
+      output_value = value;
+    }
+    else
+    {
+      throw e;
+    }
+  }
+
+  if (got_param)
+  {
+    return true;
+  }
+  else
+  {
+    RCLCPP_WARN_STREAM(logger, "Unable to get parameter: \'" << param_name << "\'. Please check YAML file");
+    return false;
+  }
+}
+
+/**
+ * Declares, reads, and validates parameters used for moveit_servo
+ * @param node Shared ptr to node that will the parameters will be declared in. Params should be defined in launch/config files
+ * @param logger Logger for outputting warnings about the parameters
+ * @param parameters The set up parameters that will be updated. After this call, they can be used to start a Servo instance
+ * @param ns Parameter namespace (as loaded in launch files). Defaults to "moveit_servo" but can be changed to allow multiple arms/instances
+ * @return true if all parameters were loaded and verified successfully, false otherwise
+ */
+bool readParameters(rclcpp::Node::SharedPtr& node, const rclcpp::Logger& logger, ServoParametersPtr& parameters, std::string ns = "moveit_servo")
```

Comment:
Reading here...  https://stackoverflow.com/a/8385731/3499467

it says SharedPtr's should be passed by const ref, so do: `const rclcpp::Logger&`

And it has some good advice for passing by reference vs value

Reply:
Looks good in most other places, changed it here

Reply:
When you pass a shared_ptr by reference you are circumventing the lifetime sharing because you are essentially passing a pointer to the shared_ptr class and therefore you are not incrementing/decrementing the counter.  If you are actually sharing the lifetime of a shared_ptr you should pass it by value (the object you are calling it on is retaining it to use it later).

As a matter of style, I generally prefer you to pass the raw pointer or a reference to the underlying object.  However, that is not the normal ros style of the const shared_ptr&.  One advantage of the const shared_ptr& is that it is generally less likely you get a nullptr but it is still possible.  IMO the best interface would be to accept a reference to the underlying object but the semantics for calling that from a shared_ptr is really annoying.  Hence the const shared_ptr& semantics.

Generally, anything that receives a pointer parameter type (including smart ones) should do nullptr checks before use.  This isn't common in ROS though, so feel free to ignore this advice.

There is a bunch of lifetime and scope management that happens in c++ that is not declarative.  This implied part of the language that is a side effect of various things that make common situations unintuitive.  There have been some efforts to help fix that like the not_null<> template but to use that we'd have to depend on yet another external library (https://github.com/microsoft/GSL).